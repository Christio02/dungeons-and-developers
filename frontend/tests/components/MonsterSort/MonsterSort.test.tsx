import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';
import SortDropdown from '../../../src/components/MonsterSort/MonsterSort.tsx';

describe('SortDropdown Component', () => {
  const mockOnSortChange = vi.fn();

  afterEach(() => {
    vi.restoreAllMocks();
    mockOnSortChange.mockReset();
  });

  it('renders correctly and matches snapshot', () => {
    const { container } = render(<SortDropdown selectedSort="name-asc" onSortChange={mockOnSortChange} />);
    expect(container).toMatchSnapshot();
  });

  it('opens and closes the dropdown menu', async () => {
    render(<SortDropdown selectedSort="name-asc" onSortChange={mockOnSortChange} />);

    const toggleButton = screen.getByText('A-Z');
    await userEvent.click(toggleButton);

    await waitFor(() => {
      expect(screen.getByText('Z-A')).toBeInTheDocument();
    });

    const button = screen.getByLabelText('close-button');
    await userEvent.click(button);

    await waitFor(() => {
      expect(screen.queryByText('Z-A')).not.toBeInTheDocument();
    });
  });

  it('calls onSortChange when a sort option is selected', async () => {
    render(<SortDropdown selectedSort="name-asc" onSortChange={mockOnSortChange} />);

    const toggleButton = screen.getByText('A-Z');
    await userEvent.click(toggleButton);

    const sortOption = screen.getByText('Difficulty: High-Low');
    await userEvent.click(sortOption);

    expect(mockOnSortChange).toHaveBeenCalledTimes(1);
    expect(mockOnSortChange).toHaveBeenCalledWith('difficulty-desc');
  });

  it('highlights the selected sort option', async () => {
    render(<SortDropdown selectedSort="difficulty-desc" onSortChange={mockOnSortChange} />);

    const toggleButton = screen.getByText('Difficulty: High-Low');
    await userEvent.click(toggleButton);

    const selectedOption = screen.queryAllByText('Difficulty: High-Low')[1];
    expect(selectedOption).toHaveClass('font-bold');
  });

  it('closes the dropdown when clicking outside', async () => {
    render(<SortDropdown selectedSort="name-asc" onSortChange={mockOnSortChange} />);

    const toggleButton = screen.getByText('A-Z');
    await userEvent.click(toggleButton);

    await waitFor(() => {
      expect(screen.getByText('Z-A')).toBeInTheDocument();
    });

    fireEvent.mouseDown(document.body); // click outside

    await waitFor(() => {
      expect(screen.queryByText('Z-A')).not.toBeInTheDocument();
    });
  });
});
