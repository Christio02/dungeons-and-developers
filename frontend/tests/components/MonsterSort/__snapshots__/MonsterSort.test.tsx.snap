// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SortDropdown Component > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="relative flex items-center"
  >
    <label
      class="mr-4 whitespace-nowrap sub-header"
    >
      Sort By:
    </label>
    <div
      class="relative"
    >
      <button
        class="px-1 rounded-lg border-2 bg-customRed sub-header text-nowrap hover:bg-transparent border-customRed hover:text-customRed transition-colors duration-200 flex items-center justify-between min-w-48 lg:min-w-60 "
      >
        A-Z
        <svg
          class="ml-3"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
